apps/www/registry/default/ui/input.tsx

```diff
@@ -1,25 +1,103 @@
 import * as React from "react"
+import { Slot } from "@radix-ui/react-slot"
+import { cva, type VariantProps } from "class-variance-authority"
 
 import { cn } from "@/lib/utils"
+import { useComposition } from "@/registry/default/hooks/use-composition"
 
+interface InputComposition {
+  Icon: typeof InputIcon
+}
+
+const iconVariants = cva("absolute top-3", {
+  variants: {
+    size: {
+      default: "h-4 w-4",
+    },
+    side: {
+      left: "left-3",
+      right: "right-3",
+    },
+  },
+  defaultVariants: {
+    size: "default",
+    side: "left",
+  },
+})
+
+export interface InputIconProps
+  extends React.HTMLAttributes<HTMLOrSVGElement>,
+    VariantProps<typeof iconVariants> {}
+
+const InputIcon = React.forwardRef<HTMLSlotElement, InputIconProps>(
+  ({ children, className, size, side }, ref) => {
+    return (
+      <Slot
+        data-icon
+        ref={ref}
+        className={cn(iconVariants({ size, side }), className)}
+      >
+        {children}
+      </Slot>
+    )
+  }
+)
+InputIcon.displayName = "InputIcon"
+
+const inputVariants = cva(
+  "flex h-10 w-full rounded-md border border-input bg-background py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50",
+  {
+    variants: {
+      composition: {
+        true: "px-10",
+        false: "px-3",
+      },
+    },
+    defaultVariants: {
+      composition: false,
+    },
+  }
+)
 export interface InputProps
   extends React.InputHTMLAttributes<HTMLInputElement> {}
 
 const Input = React.forwardRef<HTMLInputElement, InputProps>(
   ({ className, type, ...props }, ref) => {
+    return <input type={type} className={className} ref={ref} {...props} />
+  }
+)
+Input.displayName = "Input"
+
+const Root = React.forwardRef<HTMLInputElement, InputProps>(
+  ({ children, className, ...props }, ref) => {
+    const Icons = useComposition(children, InputIcon.displayName)
+
+    if (Icons.length > 0) {
```

Comment:
I’ve noticed one issue with composition. For my use case and most of the time when building other input components I’ve been playing with the `padding-x` property only on the side where the icon is added. This way, there’s no extra space on both sides from only one icon being present.

Something like this is what I thought could do the job:

```
 ...
 const Icons = useComposition(children, InputIcon.displayName);

    if (Icons.length > 0) {
      // Since we're sure that the children are InputIcon components, we can cast them to the correct type
      // to access the `side` prop
      type IconElement = React.ReactElement<InputIconProps>;

      const hasRightIcon = Icons.some(
        (icon) => (icon as IconElement).props.side === 'right',
      );
      const hasLeftIcon = Icons.some(
        (icon) =>
          (icon as IconElement).props.side === 'left' ||
          (icon as IconElement).props.side === undefined,
      );
      
      // conditionally check for composition (maybe add more variants such as left, right, true, false)
   ...

```

Reply:
Sorry for any confusion. The whole idea is to not have input padding on both sides if only one icon is present but instead only on the icon side.